cl-postgres/scram.lisp

```diff
@@ -0,0 +1,296 @@
+  (if (= 0 (search client-nonce server-nonce))
+      t
+      (error 'protocol-violation
+             :message "Client-nonce not found at beginning of server-nonce")))
+
+(defun parse-scram-server-first-response (response client-nonce &key (response-type :base64-usb8-array))
+  "Takes a server response and a client-nonce. If the server response is not in the form of an array of bytes which are encoded in base64, the response type must be specified as either :base64-string or :utf8-string. The client-nonce should be a normal utf8 string.
+It returns the server-response as a normal string, the server-provided-salt as a normal string, and the server-iterations as an integer"
+  (cond ((eq response-type :base64-usb8-array) nil)
+        ((eq response-type :utf8-string) (setf response (cl-base64:base64-string-to-usb8-array (cl-base64:string-to-base64-string response))))
+        ((eq response-type :base64-string) (setf response (cl-base64:base64-string-to-usb8-array response)))
+        (t (cerror "Invalid response type for parse-scram-server-first-response. Must be one of :base64-usb8-array, :utf8-string or :base64-string"
+                   response-type)))
+
+  (let* ((split-response (split-server-response response))
+         (server-nonce (cdr (assoc "r" split-response :test 'equal)))
+         (server-nonce-validated (validate-server-nonce server-nonce client-nonce))
+         (server-salt (cdr (assoc "s" split-response :test 'equal)))
+         (server-iterations (parse-integer (cdr (assoc "i" split-response :test 'equal))))
+         (num-of-split (length split-response)))
+    (when (not server-nonce-validated) (error 'cl-postgres-error:protocol-violation :message "Server did not validate client nonce"))
+    (when (not (= 3 num-of-split)) (error 'protocol-error
+                                          :message (format nil "There was an error in parsing the server response. Parsing had ~a results instead of 3" num-of-split)))
+    (values server-nonce server-salt server-iterations)))
+
+(defun gen-client-key (salted-password &optional (message "Client Key") (sha-method :sha256))
+  "Returns a byte array"
+  (when (stringp salted-password)
+    (setf salted-password (ironclad:ascii-string-to-byte-array salted-password)))
+  (ironclad:hmac-digest
+   (ironclad:update-hmac
+    (ironclad:make-hmac salted-password sha-method)
+    (ironclad:ascii-string-to-byte-array message))))
+
+(defun gen-stored-key (client-key)
+
+  (ironclad:digest-sequence :sha256 client-key))
+
+(defun gen-auth-message (client-initial-response server-response final-message-part1)
+  "Currently assumes all parameters are normal strings"
+  (format nil "~a,~a,~a"
+          (if (and (search "n,," client-initial-response)
+                   (= 0 (search "n,," client-initial-response)))
+              (subseq client-initial-response 3)
+              (format nil "~a" client-initial-response))
```

Comment:
Here, we format with `~A` the object `client-initial-response` which will then be once again formatted with `~A`. Is this duplicate formatting required?

Reply:
No. Agreed.